### PR TITLE
Compiler: no longer split blocks at fun call to propagate location

### PR DESCRIPTION
**Pull request by hhugo**
Trying to address https://discuss.ocaml.org/t/dune-build-vs-makefile/11394/9.

It should also help with https://github.com/ocsigen/js_of_ocaml/issues/1300
and fix https://github.com/ocsigen/js_of_ocaml/issues/747

@vouillon, the PR is not ready yet but what do you think of this approach ?

@pmwhite, this could improve your compilation time significantly.

### CODE REVIEW

**Comment by hhugo** [comment]
> Maybe we should use this instead of adding a field to every constructors of type `instr`?
> 
> ```ocaml
> type block =
>   { params : Var.t list
>   ; body : (instr * loc) list
>   ; branch : last * loc
>   }
> ```

Yeah, probably. Ill do the change and run a quick benchmark to make sure there performance profile is the same.

File: compiler/lib/parse_bytecode.ml [hunk 1]
```diff
@@ -258,31 +258,21 @@ end = struct
...
       else
         let pos =
           if after
           then loc.Location.loc_end
-          else if before
-          then loc.Location.loc_start
           else
             match event.ev_kind with
-            | Event_after _ -> loc.Location.loc_end
+            | Event_after _ when false -> loc.Location.loc_end
```
Comment (vouillon) [inline compiler/lib/parse_bytecode.ml:275]: For an event that is after an expression, the relevant location should be at the end of the expression, shouldn't it?

**Comment by hhugo** [comment]
Note that I'm still exploring the design space of this change. I'm not really happy with the mapping between debug event and instructions.

**Comment by hhugo** [comment]
Building the toplevel with sourcemaps goes from 37.6s to 28.2s

**Comment by hhugo** [comment]
> There is a mismatch between where OCaml put events, and where the JavaScript debuggers might stop. I'm not sure what to do about that.

Can you give more details ? are you just referring to the fact `debugger` is a statement ?

**Comment by hhugo** [comment]
I've cleaned up the code a little and recovered few locations that initially disappeared. I'm still trying find a better way to map event to instruction.

**Comment by hhugo** [comment]
@pmwhite, would you be able to test this and report on both compilation time and size of generated sourcemap. 
This PR both add and remove locations. I'd be curious to see if how the size of sourcemap is affected.

**Comment by vouillon** [comment]
>

**Comment by hhugo** [comment]
The PR seems to fix two issues already
- remove some completely unrelated location
- fix a compilation time issue when debug is enabled

Maybe this PR should be only do the right think to existing well understood scenario: events after a call and event before tail calls

**Comment by hhugo** [comment]
@vouillon, can I request another round of review from you ?

**Pull request merged**